base_layer/core/src/transaction.rs

```diff
@@ -84,10 +187,113 @@ pub struct TransactionKernel {
     /// Remainder of the sum of all transaction commitments. If the transaction
     /// is well formed, amounts components should sum to zero and the excess
     /// is hence a valid public key.
-    pub excess: Commitment,
+    pub excess: Option<Commitment>,
     /// The signature proving the excess is a valid public key, which signs
     /// the transaction fee.
-    pub excess_sig: Signature,
+    pub excess_sig: Option<Signature>,
 }
 
-pub type BlindingFactor = [u8; 32];
+/// Implementation of the transaction kernel
+impl TransactionKernel {
+    /// Creates an empty transaction kernel
+    pub fn empty() -> TransactionKernel {
+        TransactionKernel { features: KernelFeatures::empty(), fee: 0, lock_height: 0, excess: None, excess_sig: None }
+    }
+
+    /// Build a transaction kernel with the provided fee
+    pub fn with_fee(mut self, fee: u64) -> TransactionKernel {
+        self.fee = fee;
+        self
+    }
+
+    /// Build a transaction kernel with the provided lock height
+    pub fn with_lock_height(mut self, lock_height: u64) -> TransactionKernel {
+        self.lock_height = lock_height;
+        self
+    }
+}
+
+/// Implement the canonical hashing function for TransactionKernel for use in ordering
+impl Hashable for TransactionKernel {
+    type Hasher = Blake256;
+
+    fn hash(&self) -> Vec<u8> {
+        let mut hasher = Self::Hasher::new();
+        hasher.input(vec![self.features.bits]);
+        let mut fee_lock_height = Vec::new();
+        fee_lock_height.write_u64::<LittleEndian>(self.fee).unwrap();
+        fee_lock_height.write_u64::<LittleEndian>(self.lock_height).unwrap();
+        hasher.input(fee_lock_height);
+        if self.excess.is_some() {
+            hasher.input(self.excess.unwrap().to_bytes());
+        }
+        if self.excess_sig.is_some() {
+            hasher.input(self.excess_sig.unwrap().get_signature().to_bytes());
+        }
+        hasher.result().to_vec()
+    }
+}
+
+hashable_ord!(TransactionKernel);
+
+/// A transaction which consists of a kernel offset and an aggregate body made up of inputs, outputs and kernels.
+pub struct Transaction {
+    /// This kernel offset will be accumulated when transactions are aggregated to prevent the "subset" problem where
+    /// kernels can be linked to inputs and outputs by testing a series of subsets and see which produce valid
+    /// transactions
+    pub offset: BlindingFactor,
+    /// The constituents of a transaction which has the same structure as the body of a block.
+    pub body: AggregateBody,
+    /// reference to the Base point used in the commitments in this transaction
+    pub base: &'static Base,
+}
+
+impl Transaction {
+    /// Create a new transaction from the provided inputs, outputs, kernels and offset
+    pub fn new(
+        base: &'static Base,
```

Comment:
Phwoah.. I see why you don't like the current `HomomorphicCommitment` implementation. `Transaction` really shouldn't have a `base` field, so if this is a kludge to make it work, I definitely think we should rework the commitment struct.

Reply:
Ja, the reason it is in there is because the `Transaction` struct has to create some commitments of its own during validation (it has to create a fees commitment for example) and because creating a commitment of the form of `HomomorphicCommitment` requires a reference to the base to be provided I need to keep track of it in the `Transaction`.